heft-plugins/heft-api-extractor-plugin/src/ApiExtractorPlugin.ts

```diff
@@ -0,0 +1,246 @@
+   * `IExtractorInvokeOptions.typescriptCompilerFolder` API option. This option defaults to false.
+   */
+  useProjectTypescriptVersion?: boolean;
+}
+
+export default class ApiExtractorPlugin implements IHeftTaskPlugin {
+  private _apiExtractorPromise: Promise<typeof TApiExtractor | undefined> | undefined;
+  private _apiExtractorConfigurationFilePathPromise: Promise<string | undefined> | undefined;
+  private _apiExtractorTaskConfigurationPromise:
+    | Promise<IApiExtractorPluginConfiguration | undefined>
+    | undefined;
+
+  public apply(taskSession: IHeftTaskSession, heftConfiguration: HeftConfiguration): void {
+    taskSession.hooks.clean.tapPromise(PLUGIN_NAME, async (cleanOptions: IHeftTaskCleanHookOptions) => {
+      // Load up the configuration, but ignore if target files are missing, since we will be deleting
+      // them anyway.
+      const apiExtractorConfiguration: TApiExtractor.ExtractorConfig | undefined =
+        await this._getApiExtractorConfigurationAsync(
+          taskSession,
+          heftConfiguration,
+          /* ignoreMissingEntryPoint: */ true
+        );
+      if (apiExtractorConfiguration) {
+        await this._updateCleanOptionsAsync(cleanOptions, apiExtractorConfiguration);
+      }
+    });
+
+    taskSession.hooks.run.tapPromise(PLUGIN_NAME, async (runOptions: IHeftTaskRunHookOptions) => {
+      const apiExtractor: typeof TApiExtractor | undefined = await this._getApiExtractorAsync(
+        taskSession,
+        heftConfiguration
+      );
+      const apiExtractorConfiguration: TApiExtractor.ExtractorConfig | undefined =
+        await this._getApiExtractorConfigurationAsync(taskSession, heftConfiguration);
+      if (apiExtractor && apiExtractorConfiguration) {
+        await this._runApiExtractorAsync(
+          taskSession,
+          heftConfiguration,
+          runOptions,
+          apiExtractor,
+          apiExtractorConfiguration
+        );
+      }
+    });
+  }
+
+  private async _getApiExtractorConfigurationFilePathAsync(
+    heftConfiguration: HeftConfiguration
+  ): Promise<string | undefined> {
+    if (!this._apiExtractorConfigurationFilePathPromise) {
+      this._apiExtractorConfigurationFilePathPromise =
+        heftConfiguration.rigConfig.tryResolveConfigFilePathAsync(CONFIG_FILE_LOCATION);
+    }
+    return await this._apiExtractorConfigurationFilePathPromise;
+  }
+
+  private async _getApiExtractorConfigurationAsync(
+    taskSession: IHeftTaskSession,
+    heftConfiguration: HeftConfiguration,
+    ignoreMissingEntryPoint?: boolean
+  ): Promise<TApiExtractor.ExtractorConfig | undefined> {
+    const configObjectFullPath: string | undefined = await this._getApiExtractorConfigurationFilePathAsync(
+      heftConfiguration
+    );
+    if (!configObjectFullPath) {
+      return undefined;
+    }
+
+    const apiExtractor: typeof TApiExtractor = (await this._getApiExtractorAsync(
+      taskSession,
+      heftConfiguration
+    ))!;
```

Comment:
WRT this non-null assertion - do we know that we'll be able to resolve API Extractor here?

Reply:
Yes and no. If the call succeeds, we will have API Extractor. If it fails, it will throw a helpful error message about being unable to resolve API extractor from the project/rig. I'll split this up better to make this expectation clear